core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala

```diff
@@ -442,31 +442,43 @@ class CassandraJournal(cfg: Config) extends AsyncWriteJournal
     }
 
     /**
-     * Deletes the by inserting into the metadata table deleted_to
+     * Deletes the events by inserting into the metadata table deleted_to
      * and physically deletes the rows.
      *
      * The Future completes once the logical delete is complete as
      * that is when future reads won't see the events.
      */
-    def logicalDeleteWithAsyncPhysicalDelete(lowestSequenceNr: Long, highestSequenceNr: Long): Future[Done] = {
-      val lowestPartition = partitionNr(lowestSequenceNr, config.targetPartitionSize)
+    def logicalDeleteWithAsyncPhysicalDelete(highestDeletedSequenceNumber: Long, highestSequenceNr: Long): Future[Done] = {
+      val lowestPartition = partitionNr(highestDeletedSequenceNumber, config.targetPartitionSize)
       val toSeqNr = math.min(toSequenceNr, highestSequenceNr)
       val highestPartition = partitionNr(toSeqNr, config.targetPartitionSize) + 1 // may have been moved to the next partition
-      val boundInsertDeletedTo = preparedInsertDeletedTo.map(_.bind(persistenceId, toSeqNr: JLong))
-      val logicalDelete = boundInsertDeletedTo.flatMap(session.executeWrite)
-      // Done async as thw plugin won't see the events once the logical delete is complete
+      val logicalDelete =
+        if (toSeqNr <= highestDeletedSequenceNumber) {
+          // already deleted same or higher sequence number, don't update highestDeletedSequenceNumber,
+          // but perform the physical delete (again), may be a retry request since to may fail
+          Future.successful(())
+        } else {
+          val boundInsertDeletedTo = preparedInsertDeletedTo.map(_.bind(persistenceId, toSeqNr: JLong))
+          boundInsertDeletedTo.flatMap(session.executeWrite)
+        }
+      // Done async as the plugin won't see the events once the logical delete is complete
       logicalDelete.foreach(_ => physicalDelete(lowestPartition, highestPartition, toSeqNr))
```

Comment:
I think the reason was that the physical delete could run in the background without affecting the response time of the PersistentActor.

On the other hand I guess there is no reason for a PersistentActor to "wait" or depend on how long time it takes for the delete to complete, so one could say that the entire delete is running in the background.

One problem with adding that to the response time might be that it could trigger the CircuitBreaker in Akka.

Let's see if there are other opinions when this PR is reviewed.

Reply:
This makes sense so that if a user actor gets a delete success they are definitely fully deleted.
Also completing the Future earlier only effects when the user actor gets the delete success it doesn't block further writes so it isn't really optimising anything unless the user is doing a side effect on delete success.

Maybe under a diff PR or commit?

Reply:
Please :)

Reply:
sure, I'll change that tomorrow morning